### DISCUSSION TITLE
Could use some help demuxing SFD files

### DISCUSSION THREAD

**Discussion by throwaway1908** [ThisKwasior/CryTools#5] [Q&A] [answered]
Hi, I was wondering if I could get some help demuxing some SFD files? The files I am trying to demux are from the Xbox version of Sonic Heroes. I am not exactly sure what command line I should enter to extract the mpg/adx files from it. If you are able to help, that will be much appreciated.

**Comment by ThisKwasior** [comment, marked as answer]
You can demux SFDs with FFmpeg no problem. Just checked on one cutscene from the Xbox version.
As for audio, some files are ADX and some are AIX, keep that in mind.

Demuxing video:
`ffmpeg -loglevel quiet -i E0000_XD.sfd -an -c copy E0000_XD.m2v`

Demuxing audio (4 streams):
`ffmpeg -loglevel quiet -i "E0000_XD.sfd" -vn -c copy -map 0:a:0 E0000_XD_a0.adx`
`ffmpeg -loglevel quiet -i "E0000_XD.sfd" -vn -c copy -map 0:a:1 E0000_XD_a1.adx`
`ffmpeg -loglevel quiet -i "E0000_XD.sfd" -vn -c copy -map 0:a:2 E0000_XD_a2.adx`
`ffmpeg -loglevel quiet -i "E0000_XD.sfd" -vn -c copy -map 0:a:3 E0000_XD_a3.adx`

If there's `AIXF` at the start of the file, then it's AIX so you can rename the file or play it with ffplay like this:
`ffplay -f aix -i E0000_XD_a1.adx`

**Reply by throwaway1908** [reply]
I have tried with ffmpeg before, but it gives a glitchy mess as a result or weird saturated colors as shown in the attached pic. 
![mpc-hc64_wsLqQS2vVh](https://github.com/ThisKwasior/CryTools/assets/60544658/00fb8489-7d55-4abf-9249-ee59676f74a7)

**Reply by ThisKwasior** [reply]
This isn't related to demuxing, but TMPGEnc 11-bit DC prediction (weird glitches) and Criware's color space compensation (saturated colors).

Glitches can be fixed with re-encoding using FFmpeg.

To fix these overexposed videos, you need to use my `csc_video_conv` program.

It goes as follows:
1. First, convert the video to raw YUV stream with FFmpeg
`ffmpeg -i E0000_XD.sfd E0000_XD.yuv`
2. Use my csc converter with the yuv stream and original sfd as parameters. It will create `E0000_XD_converted.yuv`
`csc_video_conv.exe E0000_XD.yuv E0000_XD.sfd`
3. Convert the fixed yuv stream to video, here I'm converting it to MP4.
`ffmpeg -f rawvideo -pix_fmt yuv420p -s 720x480 -framerate 29.97 -i E0000_XD_converted.yuv E0000_XD_converted.mp4`

Keep in mind it's an approximation and not 1:1 to original stream, but looks good enough.

**Reply by throwaway1908** [reply]
Would I be able to get a download link to your csc_video_conv program? And thanks so much for all your help, it is greatly appreciated.

**Reply by ThisKwasior** [reply]
Sure thing, Here's the up-to-date build from a minute ago.
[CryTools-2023.zip](https://github.com/ThisKwasior/CryTools/files/13461601/CryTools-2023.zip)
Let me know if everything works

**Reply by throwaway1908** [reply]
Yes everything works! Thank you so much.
